bot/exts/filtering/_filter_lists/filter_list.py

```diff
@@ -0,0 +1,309 @@
+            if not filter_.validations:
+                if default_answer and await filter_.triggered_on(ctx):
+                    relevant_filters.append(filter_)
+            else:
+                passed, failed = filter_.validations.evaluate(ctx)
+                if not failed and failed_by_default < passed:
+                    if await filter_.triggered_on(ctx):
+                        relevant_filters.append(filter_)
+
+        if ctx.event == Event.MESSAGE_EDIT and ctx.message and self.list_type == ListType.DENY:
+            previously_triggered = ctx.message_cache.get_message_metadata(ctx.message.id)
+            # The message might not be cached.
+            if previously_triggered:
+                ignore_filters = previously_triggered[self]
+                # This updates the cache. Some filters are ignored, but they're necessary if there's another edit.
+                previously_triggered[self] = relevant_filters
+                relevant_filters = [filter_ for filter_ in relevant_filters if filter_ not in ignore_filters]
+        return relevant_filters
+
+    def default(self, setting_name: str) -> Any:
+        """Get the default value of a specific setting."""
+        missing = object()
+        value = self.defaults.actions.get_setting(setting_name, missing)
+        if value is missing:
+            value = self.defaults.validations.get_setting(setting_name, missing)
+            if value is missing:
+                raise ValueError(f"Couldn't find a setting named {setting_name!r}.")
+        return value
+
+    def merge_actions(self, filters: list[Filter]) -> ActionSettings | None:
+        """
+        Merge the settings of the given filters, with the list's defaults as fallback.
+
+        If `merge_default` is True, include it in the merge instead of using it as a fallback.
+        """
+        if not filters:  # Nothing to action.
+            return None
+        try:
+            return reduce(
+                or_, (filter_.actions or self.defaults.actions for filter_ in filters)
+            ).fallback_to(self.defaults.actions)
+        except TypeError:
+            # The sequence fed to reduce is empty, meaning none of the filters have actions,
+            # meaning they all use the defaults.
+            return self.defaults.actions
+
+    @staticmethod
+    def format_messages(triggers: list[Filter], *, expand_single_filter: bool = True) -> list[str]:
+        """Convert the filters into strings that can be added to the alert embed."""
+        if len(triggers) == 1 and expand_single_filter:
+            message = f"#{triggers[0].id} (`{triggers[0].content}`)"
+            if triggers[0].description:
+                message += f" - {triggers[0].description}"
+            messages = [message]
+        else:
+            messages = [f"{filter_.id} (`{filter_.content}`)" for filter_ in triggers]
+        return messages
+
+    def __hash__(self):
+        return hash(id(self))
```

Comment:
Why are we hashing memory identity of the instance here? Shouldn't the instance `id` be used instead?

Reply:
I tried to make this correct beyond my current usage of it. For example, if I use AtomicList objects as dict keys, I don't want to assume what else will be put in that dict.